types/openebs.go

```diff
@@ -0,0 +1,234 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// OpenEBS defines the intent to get
+// OpenEBS deployed/updated on a Kubernetes setup
+type OpenEBS struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec OpenEBSSpec `json:"spec"`
+
+	Status OpenEBSStatus `json:"status"`
+}
+
+// OpenEBSSpec defines the specifications that determines
+// what OpenEBS (e.g. version, components, etc)
+// get deployed on a Kubernetes setup
+type OpenEBSSpec struct {
+	Version                    string            `json:"version"`
+	DefaultStoragePath         string            `json:"defaultStoragePath"`
+	CreateDefaultStorageConfig string            `json:"createDefaultStorageConfig"`
+	ImagePrefix                string            `json:"imagePrefix"`
+	ImagePullPolicy            corev1.PullPolicy `json:"imagePullPolicy"`
+	Components                 `json:",inline"`
+}
+
+// Components stores all the OpenEBS components.
+type Components struct {
+	APIServer        *APIServer        `json:"apiServer"`
+	Provisioner      *Provisioner      `json:"provisioner"`
+	LocalProvisioner *LocalProvisioner `json:"localProvisioner"`
+	SnapshotOperator *SnapshotOperator `json:"snapshotOperator"`
+	AdmissionServer  *AdmissionServer  `json:"admissionServer"`
+	NDM              *NDM              `json:"ndm"`
+	NDMOperator      *NDMOperator      `json:"ndmOperator"`
+	Jiva             *Jiva             `json:"jiva"`
+	Cstor            *Cstor            `json:"cstor"`
+	Helper           *Helper           `json:"helper"`
+	Policies         *Policies         `json:"policies"`
+	Analytics        *Analytics        `json:"analytics"`
+}
+
+// Helper stores the details of helper containers used by OpenEBS
+type Helper struct {
+	Container `json:",inline"`
+}
+
+// Policies stores the details of various policies being supported
+// by OpenEBS i.e. Monitoring.
+type Policies struct {
+	Monitoring *Monitoring `json:"monitoring"`
+}
+
+// Monitoring contains the details of monitoring container
+type Monitoring struct {
+	ComponentCommonConfig `json:",inline"`
+	Container             `json:",inline"`
+}
+
+// Analytics contains the configuration of analytics being used
+// in OpenEBS
+type Analytics struct {
+	Enabled      string `json:"enabled"`
+	PingInterval string `json:"pingInterval"`
+}
+
+// ComponentCommonConfig stores the configuration which are common/same
+// for most of the components.
+type ComponentCommonConfig struct {
+	Enabled      string              `json:"enabled"`
+	Replicas     *int32              `json:"replicas"`
+	NodeSelector map[string]string   `json:"nodeSelector"`
+	Tolerations  []corev1.Toleration `json:"tolerations"`
+	Affinity     *corev1.Affinity    `json:"affinity"`
+}
+
+// APIServer stores the configuration of maya-apiserver
+type APIServer struct {
+	ComponentCommonConfig `json:",inline"`
+	Container             `json:",inline"`
+	Sparse                *SparsePools `json:"sparse"`
```

Comment:
I am not too sure here. We might just want to make use of `SparsePools` vs. `*SparsePools`
Unless we are mutating the fields of SparsePools with some methods.